### DISCUSSION TITLE
World Package 1.5.0버전이 업데이트 되었습니다

### DISCUSSION THREAD

**Discussion by Zepeto-TechEvangelist** [naverz/zepeto-studio-kor#989] [1. Announcements]
안녕하세요, 제페토 월드 크리에이터 여러분
World Package 1.5.0 버전이 업데이트 되어서 안내드립니다.
저희 제페토 팀은 창의적인 월드를 쉽게 개발할수있는 환경을 제공하기 위해 많은 노력을 기울이고 있습니다.

향후 로드맵 역시 공개 준비 중에 있으니, 많은 기대와 응원 부탁 드립니다. 
여러분의 제보와 건의, 창작 활동에 진심으로 감사 드립니다. 

### 안내
[패키지 업데이트 방법](https://studio.zepeto.me/kr/guides/updating-package)
[1.5.0 릴리즈 노트 보기](https://studio.zepeto.me/kr/guides/world-150)

<hr/>

### 신규 추가 기능

- 기존에는 제페토 앱의 Quality 설정을 따르도록 되어있으나 ProjectSetting > Quality 설정이 개별 월드에 적용되도록 처리 https://github.com/naverz/zepeto-studio-kor/issues/751
   - 이에 따라 모바일 Quality 설정을 아래와 같이 High 이상으로 설정할 경우 발열이 발생할 수 있습니다.
<img width="1073" alt="스크린샷 2022-07-12 오전 11 19 15" src="https://user-images.githubusercontent.com/100118650/178394487-7e7c25b6-100f-4a86-8e6b-8d57a50bb9d2.png">

- CharacterController - StateMachine에 접근하여, Transition time 조정 가능하도록 처리 https://github.com/naverz/zepeto-studio-kor/discussions/653
- Invoke(string methodName, float time) 지원
- OnDrawGizmos 지원 https://github.com/naverz/zepeto-studio-kor/discussions/872
- IEnumerable Type에 대해 for of 문이 가능하도록 지원
- PostProcessing GetSetting / TryGetSettings 지원 https://github.com/naverz/zepeto-studio-kor/discussions/882
- Leaderboard - SetScore API 사용 시, ExtraInfo (string) 파라메터를 제어할 수 있도록 제공
   - ExtraInfo : 각 Rank(랭킹정보) 단위로 활용할 수 있는 추가 데이터 타입으로, SetScore API를 통해 유효한 Ranking이 리더보드에 기록될 때 함께 갱신 할 수 있습니다.
   - ex) 플레이어가 승리할 때 착용 중인 ItemID 정보를 랭킹과 함께 기록(갱신) 가능

<hr/>

### IWP 관련 변경 사항

**[IWP 개선 사항]**
- Unity 에디터 테스트 및 QR코드를 통한 모바일 테스트 시 비소모성 IWP 상품을 한 번 구매하면 이후에 테스트할 때는 재구매가 불가능한 부분을 개선
   - 에디터 테스트 시 비소모성 IWP 상품 구매 후 인스펙터에서 Clear 버튼 노출로 재구매가 가능한 환경 제공 
   - QR코드를 통한 모바일 테스트 시에는 비소모성 IWP 상품 구매 후, QR 모바일 재접속 시 재구매 가능
- IWP 상품이 있는 월드를 출시했고, 새 버전 추가를 하지 않은 상태에서 유니티 에디터 테스트 및 QR코드를 통한 모바일 테스트시 IWP 상품이 구매가 불가능한 부분을 개선
- 클라이언트 단에서 ZEM 구매 실패가 발생시, 실제 ZEM 차감이 안되도록 개선

**[ IWP 정책 변경 사항 ]**
- 서버 코드에서 IWP onPurchased()를 구현시, 실패 혹은 예외 처리에 대해서 반드시 로직을 구현해야 합니다.
   - IWP 상품 구매시 아이템 지급을 못 받을 경우 에러를 발생 시켜서,  ZEM 차감만 되는 상황이 발생하지 않도록 아래 예제를 참고해서 로직을 구현하여 주세요.
   - 실패 혹은 예외 처리 로직 구현이 되지 않아, 정상적으로 동작이 되지 않는 사례에 대해서는 월드 심사시 리젝 처리 됩니다.

```typescript
async onPurchased(client: SandboxPlayer, receipt: IReceiptMessage) {
        
    const storage = client.loadDataStorage();
    const itemStr = await storage.get('items') as string;
    const items = JSON.parse(itemStr) as string[] || [];
    const hasItem = items.find(item => item === receipt.itemId);

    // 이미 가지고 있는 상황에 대한 예외처리
    if (hasItem) {
        throw new Error('already purchased'); // 사용자의 젬이 차감되지 않습니다.
    }

    if (!hasItem) {
        items.push(receipt.itemId);
        await storage.set('items', JSON.stringify(items));
    }

    // 로직에 문제가 없이 onPurchased 가 종료되면, 사용자의 젬이 정상적 차감됩니다.
}

```

<hr/>

### 개선된 기능

- 런타임에 AddComponent 한 이후 GetComponent가 동작하지 않는 문제 수정 https://github.com/naverz/zepeto-studio-kor/discussions/528
- 패키지 업데이트 이후 Zepeto World Multiplay 컴포넌트가 Missing 되는 문제 수정 https://github.com/naverz/zepeto-studio-kor/discussions/874
- GameObject 타입 변수의 null 비교 가능하도록 수정 https://github.com/naverz/zepeto-studio-kor/discussions/466
- 런타임에 제페토 스크립트 추가 시 Start() 이벤트가 누락되는 문제 수정 https://github.com/naverz/zepeto-studio-kor/issues/773
- iOS에서 특정 UI 버튼을 길게 눌러야 반응하는 현상 수정 https://github.com/naverz/zepeto-studio-kor/discussions/369
- 간헐적으로 인스펙터창에서 public 변수가 사리지고 ArgumentNullException 에러 로그가 출력되던 문제 수정 https://github.com/naverz/zepeto-studio-kor/discussions/631, https://github.com/naverz/zepeto-studio-kor/discussions/908
- GetComponentsInChildren 함수를 이용했을 때 배열의 인자가 undefined로 반환되는 문제 수정 https://github.com/naverz/zepeto-studio-kor/issues/819
- 프리팹을 Scene에 복제했을 때, 복제된 프리팹에서 프로퍼티가 변경되지 않는 문제 수정 https://github.com/naverz/zepeto-studio-kor/issues/771,  https://github.com/naverz/zepeto-studio-kor/discussions/662
- 레이어 설정이 적용되지 않는 문제 수정 https://github.com/naverz/zepeto-studio-kor/issues/799
- Exception 오류 메세지 오탈자 수정 https://github.com/naverz/zepeto-studio-kor/issues/744
- Windows 환경에서 빌드 된 월드를 모바일에서 실행 시 Resource.Load가 정상적으로 동작하지 않는 문제 수정
- GameObject의 Active 상태 전환 시 Update(), LateUpdate()가 재시작하지 않는 문제 수정 https://github.com/naverz/zepeto-studio-kor/issues/773
- ProjectSetting > CollisionLayer 설정 적용이 되지 않는 문제 수정 https://github.com/naverz/zepeto-studio-kor/issues/799
- ZepetoPlayers.instance.GetPlayerWithUserId API 사용시 One Line으로 코드 작성하면 GetComponent가 안되는 문제 수정 https://github.com/naverz/zepeto-studio-kor/issues/808
- CompareTag 가 안드로이드에서 동작하지 않는 문제 수정 https://github.com/naverz/zepeto-studio-kor/discussions/465
- Update 에서 컴포넌트가 추가되는 경우 LateUpdate 가 먼저 발생하지 않는 문제 수정

**Comment by Cha-SeongHun** [comment]
https://studio.zepeto.me/kr/guides/handling-iwp-event(server)
링크의 IWP Server handler 가이드에 변경된 정책에 대한 내용을 찾아볼 수 없습니다.
가이드에도 이 부분 추가해주시면 감사하겠습니다

**Reply by Zepeto-TechEvangelist** [reply]
안녕하세요, 관련 가이드 문서 한글 버전 우선 업데이트 완료되었습니다.
감사합니다.
